src/high-level/matrix.lisp

```diff
@@ -248,15 +248,42 @@ ELEMENT-TYPE, CAST, COPY-TENSOR, DEEP-COPY-TENSOR, TREF, SETF TREF)"
 
 ;; Specific constructors
 
-(defgeneric random-unitary (shape &key type)
-  (:documentation "Generate a uniformly random element of U(n).")
-  (:method (shape &key (type 'double-float))
-    (policy-cond:with-expectations (> speed safety)
-        ((assertion (square-shape-p shape)))
-      (multiple-value-bind (q r) (qr (rand shape :type type :distribution #'alexandria:gaussian-random))
-        (let ((d (diag r)))
-          (setf d (cl:map 'list (lambda (di) (/ di (sqrt (* di (conjugate di))))) d))
-          (@ q (funcall #'from-diag d)))))))
+(defgeneric fill-with-random-normal! (matrix)
+  (:method ((matrix matrix/single-float))
+    (into! (lambda (i j)
+             (declare (ignore i j))
+             (coerce (alexandria:gaussian-random) 'single-float))
+           matrix))
+  (:method ((matrix matrix/double-float))
+    (into! (lambda (i j)
+             (declare (ignore i j))
+             (values (alexandria:gaussian-random)))
+           matrix))
+  (:method ((matrix matrix/complex-single-float))
+    (into! (lambda (i j)
+             (declare (ignore i j))
+             (multiple-value-bind (re im) (alexandria:gaussian-random)
+               (complex (coerce re 'single-float)
+                        (coerce im 'single-float))))
+           matrix))
+  (:method ((matrix matrix/complex-double-float))
+    (into! (lambda (i j)
+             (declare (ignore i j))
+             (multiple-value-call #'complex (alexandria:gaussian-random)))
+           matrix)))
+
+(defun random-normal (shape &key (type *default-tensor-type*))
+  "Produce a matrix of entries which are normally distributed (mean 0, variance 1). Complex types will have their real and imaginary parts normally distributed."
+  (fill-with-random-normal! (zeros shape :type type)))
+
+(defun random-unitary (shape &key (type `(complex ,*default-tensor-type*)))
+  "Generate a uniformly random element of U(n)."
+  (policy-cond:with-expectations (> speed safety)
+      ((assertion (square-shape-p shape)))
+    (multiple-value-bind (q r) (qr (random-normal shape :type type))
+      (let ((d (diag r)))
+        (setf d (cl:map 'list (lambda (di) (/ di (sqrt (* di (conjugate di))))) d))
```

Comment:
This is the right thing to do, but the extra scaling might not be obvious to a reader. I suggest to drop a link to a reference (e.g. https://arxiv.org/pdf/math-ph/0609050.pdf)